### DISCUSSION TITLE
Python API

### DISCUSSION THREAD

**Discussion by demberto** [pyinstaller/pyinstaller#7636] [Features & Development]
Spec files already contain Python code. Why not have an API for users to be able to make "build" scripts, which provide a precise control over what files are collected and dropped, while having complete access to Python itself?

I am using PyInstaller for a project of mine using Gtk. While the hooks and built executable are good, I'd like to have a precise control over what icons are included (I only use scalable icons). I'd also like to exclude all UCRT and `api-ms-` (APISet) DLLs entirely, as I feel runtimes should be shared.

I am sure this must be asked somewhere in issues, but the keyword "API" returns a shitload of irrelevant results.

I also found PyInstaller type stubs on PyPI, does this mean there's an internal API already?

**Comment by bwoodsend** [comment]
I don't understand the difference between a spec file and what you're asking for? You can already stick arbitrary code after the `a = Analysis()` section to filter icons from `a.datas` or runtime libraries from `a.binaries`.

**Reply by demberto** [reply]
My purpose aside, just asking, a spec file is ran by PyInstaller rather than Python itself. This means whatever code you add into the spec file can impact the locals/globals/imports (or just code) added by PyInstaller itself.

Redefinitions, change of environment variables etc. can happen which can cause weird bugs.
